File: tests/fixtures/deadcode/scala/Fixture.scala
package deadcode.fixture

import akka.actor.Actor
import org.junit.Test
import org.scalatest.funsuite.AnyFunSuite
import play.api.mvc.InjectedController

trait Task {
  def run(): Int
}

class DefaultTask extends Task {
  override def run(): Int = Helpers.directlyUsedHelper()
}

object Helpers {
  def unusedCleanupCandidate(): Int = 42

  def directlyUsedHelper(): Int = 7

  def generatedExcludedHelper(): Int = 13
}

class PublicService {
  def status(): String = "ok"
}

class JobEndpoint extends InjectedController {
  def handle = Action { "handled" }
}

class WorkerActor extends Actor {
  override def receive: Receive = {
    case "run" => sender() ! Helpers.directlyUsedHelper()
  }
}

class FixtureSuite extends AnyFunSuite {
  @Test
  def exercisedByTestRunner(): Unit = {}
}

object ScriptMain extends App {
  Helpers.directlyUsedHelper()
}

object Main {
  val dynamicMethodName = "run"

  def main(args: Array[String]): Unit = {
    val task: Task = () => Helpers.directlyUsedHelper()
    task.run()
    new PublicService().status()
    new JobEndpoint().handle()
  }
}

File: tests/fixtures/deadcode/rust/fixture.rs
trait Task {
    fn run(&self) -> i32;
}

struct ReportTask;

impl Task for ReportTask {
    fn run(&self) -> i32 {
        directly_used_helper()
    }
}

fn unused_cleanup_candidate() -> i32 {
    42
}

fn directly_used_helper() -> i32 {
    7
}

pub fn public_status() -> &'static str {
    "ok"
}

fn registered_handler() -> i32 {
    3
}

fn generated_excluded_helper() -> i32 {
    13
}

const DYNAMIC_FUNCTION_NAME: &str = "registered_handler";

fn main() {
    let task: Box<dyn Task> = Box::new(ReportTask);
    let handler: fn() -> i32 = registered_handler;
    let _ = task.run() + handler();
    let _ = public_status();
}
